### PR TITLE
Separate the Functions validation and transforamtion

### LINKED ISSUE #9180: [Function] Separate the Functions validation and transformation
**Motivation**

When creating a function, we will do the validation for the function package and transform the FunctionConfig to the FunctionDetail in the same method.
We introduced package management for saving the function packages, and want to reduce the duplicated validation when creating a function with a package name.
So I suggest we need to separate the validation steps and transformation step to make the method more clear.

### PR DESCRIPTION
**Pull request by zymap**
---

Fixes #9180

*Motivation*

When creating a function, we will do the validation for the
function package and transform the FunctionConfig to the
FunctionDetail in the same method.
We introduced package management for saving the function
packages, and want to reduce the duplicated validation when
creating a function with a package name.
So I suggest we need to separate the validation steps and
transformation step to make the method more clear.

### Verifying this change

Because we only refine the code and doesn't change the logic, so we need to make sure the existent CI passed.

- [x] Make sure that the change passes the CI checks.

### CODE REVIEW

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java [hunk 1]
```diff
@@ -941,4 +948,16 @@ public static FunctionConfig validateUpdate(FunctionConfig existingConfig, Funct
         }
         return mergedConfig;
     }
+
+    public static boolean isPackageNameStyle(String functionPkgUrl) {
```
Comment (sijie) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java:952]:
```suggestion
    public static boolean isPulsarPackageUrl(String functionPkgUrl) {
```

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/FunctionsImpl.java [hunk 2]
```diff
@@ -89,101 +90,31 @@ public void registerFunction(final String tenant,
...
-                    functionName, clientRole, ComponentTypeUtils.toString(componentType));
-            throw new RestException(Response.Status.UNAUTHORIZED, "client is not authorize to perform operation");
-        } catch (PulsarAdminException.NotFoundException e) {
-            log.error("{}/{}/{} Tenant {} does not exist", tenant, namespace, functionName, tenant);
-            throw new RestException(Response.Status.BAD_REQUEST, "Tenant does not exist");
-        } catch (PulsarAdminException e) {
-            log.error("{}/{}/{} Issues getting tenant data", tenant, namespace, functionName, e);
-            throw new RestException(Response.Status.INTERNAL_SERVER_ERROR, e.getMessage());
-        }
+        requestParamsPreCheck(tenant, namespace, functionName, functionConfig);
```
Comment (sijie) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/FunctionsImpl.java:93]:
```suggestion
        preCheckRequestParams(tenant, namespace, functionName, functionConfig);
```

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/FunctionsImpl.java [hunk 3]
```diff
@@ -722,20 +581,101 @@ public void updateFunctionOnWorkerLeader(final String tenant,
...
+        if (namespace == null) {
+            throw new RestException(Response.Status.BAD_REQUEST, "Namespace is not provided");
+        }
+        if (functionName == null) {
+            throw new RestException(Response.Status.BAD_REQUEST, "Function name is not provided");
+        }
+    }
+
+    // validate the operating tenant exists and has proper permissions to operate
+    private void validateFunctionNamespaceOperation(String tenant, String namespace, String functionName,
```
Comment (sijie) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/FunctionsImpl.java:604]: You reverse the order of the original code. The original code checks `isAuhtorizedRole` first. Please don't change any code when you refactor code.

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/FunctionsImpl.java [hunk 4]
```diff
@@ -318,57 +227,16 @@ public void updateFunction(final String tenant,
...
-                            mergedConfig, componentPackageFile);
-                    if (!isFunctionCodeBuiltin(functionDetails) && (componentPackageFile == null || fileDetail == null)) {
-                        throw new IllegalArgumentException(ComponentTypeUtils.toString(componentType) + " Package is not provided");
-                    }
-                } else {
-
-                    componentPackageFile = FunctionCommon.createPkgTempFile();
-                    componentPackageFile.deleteOnExit();
-                    WorkerUtils.downloadFromBookkeeper(worker().getDlogNamespace(), componentPackageFile, existingComponent.getPackageLocation().getPackagePath());
+        if (FunctionConfigUtils.isPackageNameStyle(functionPkgUrl)) {
```
Comment (sijie) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/FunctionsImpl.java:230]: In the original code, there is a logic verifying `existingComponent`. After refactor, the code doesn't exist there. It is very hard for reviewers to review this code and understand the change.

**Review by sijie — CHANGES_REQUESTED** [review]

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/FunctionsImpl.java [hunk 5]
```diff
@@ -754,17 +721,19 @@ public void updateFunctionOnWorkerLeader(final String tenant,
...
+        return clsLoader;
     }
 
-    private File downloadPackageFile(String packageName) throws IOException, PulsarAdminException {
-        File file = Files.createTempFile("function", ".tmp").toFile();
-        worker().getBrokerAdmin().packages().download(packageName, file.toString());
-        return file;
+    private ImmutablePair<String, String> getTypeClassName(String functionPackageUrl) {
+        try {
+            Map<String, String> properties = worker().getBrokerAdmin().packages()
```
Comment (jerrypeng) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/FunctionsImpl.java:729]: I don't understand what we are doing here.  We are not using pulsar admin to upload packages.

Comment (zymap) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/FunctionsImpl.java:729]: We introduce [package management](https://docs.google.com/document/d/1FDSNhc8YB1yUMFBUrx8p-1nH1s5Kftp04PkFKWec-2U/edit) to manage function jars. 
We used to get the FunctionTypes by loading the jar and get it from the class. If we want to use a jar from the package management, we want to get it directly from the metadata and avoid download the jar at the broker.

**Comment by zymap** [comment]
Because this has many CI failed. I will close this PR and reopen it when it is ready.

**Pull request closed without merging**